### PR TITLE
#97 Convert Outpatient

### PR DESCRIPTION
**Pull request by catherineholland1**
Conversion of scripts:
B03 Process Outpatients (SMR00) extract 
B03X Outpatient_tests

tried to implement a more efficient way of line 117 - cost in correct month but it would not seem to work as 2 `map` would be required

### CODE REVIEW

**Comment by catherineholland1** [comment]
removed the dates from the summary - they seem to be throwing off the `pivot_longer` as all variables need to be the same type. Any ideas on this, or crucial to keep in?

**Comment by Jennit07** [comment]
@Jennit07 to pick this up

File: Production_scripts/B03_Process_Outpatients_(SMR00)_extract.R [hunk 1]
```diff
@@ -0,0 +1,245 @@
...
+# .rds file
+readr::write_rds(outfile,
+  paste0(
+    get_year_dir(year = latest_year),
+    "/outpatient_for_source-20",
+    latest_year, ".zsav"
+  ),
+  compress = "gz"
+)
+
```
Comment (Jennit07) [inline Production_scripts/B03_Process_Outpatients_(SMR00)_extract.R:245]: Use `get_source_extract_path` function for saving
```suggestion
# Save as zsav file
outfile %>%
  readr::write_rds(get_source_extract_path(year, "Outpatient", ext = "zsav"))

# Save as rds file
outfile %>%
  readr::write_rds(get_source_extract_path(year, "Outpatients", ext = "rds"))
```

File: R/create_outpatient_extract_flags.R [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+    create_demog_test_flags(postcode = postcode) %>%
+    # create HB flags
+    create_hb_test_flags(.data$hbtreatcode) %>%
+    # replace missing hb with 0
+    mutate(across(starts_with("NHS_"), ~ replace_na(.x, 0))) %>%
+    # create HB cost flags
+    create_hb_costs_test_flags(.data$cost_total_net) %>%
+    # replace missing hb costs with 0
+    mutate(across(starts_with("NHS_") & ends_with("_cost"), ~ replace_na(.x, 0)))
+}
```
Comment (Jennit07) [inline R/create_outpatient_extract_flags.R:21]: Split these onto outpatients_test branch

File: R/produce_outpatient_extract_test.R [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+        NHS_Lanarkshire_cost = sum(NHS_Lanarkshire_cost)
+      ) %>%
+      tidyr::pivot_longer(
+        cols = everything(),
+        names_to = "measure",
+        values_to = "value",
+        values_transform = list(value = as.numeric)
+      )
+  }
+}
```
Comment (Jennit07) [inline R/produce_outpatient_extract_test.R:184]: Split these onto outpatients_tests branch

**Review by Jennit07 — DISMISSED** [review]
Made some changes to the formatting of the script. Split into sections, created new function create_monthly_costs and split tests onto new branch

**Comment by Jennit07** [comment]
@Moohan ready for re-review

File: Production_scripts/B03_Process_Outpatients_(SMR00)_extract.R [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+# Version of R - 3.6.1
+# Input - episode extract csv file
+# Description - Process Outpatients extract
+#####################################################
+
+# Load packages
+library(dplyr)
+library(tidyr)
+library(createslf)
+library(vroom)
```
[the remark this answers is not in the document]
Comment (Moohan) [reply Production_scripts/B03_Process_Outpatients_(SMR00)_extract.R:15]: I think you're using vroom for the `cols_` functions which are also in readr. 

See here for the differences between vroom and readr: [https://www.tidyverse.org/blog/2021/11/readr-2-1-0-lazy/](https://www.tidyverse.org/blog/2021/11/readr-2-1-0-lazy/) basically vroom was written to be faster but was eventually included in readr. The default readr behaviour (in the latest version) is to use `lazy = FALSE` as it always has, vroom type usage would be `lazy = TRUE` I think for our case it doesn't matter as we're almost always using the entire file (and if we're not we probably should trim our extract!).

Comment (Moohan) [inline Production_scripts/B03_Process_Outpatients_(SMR00)_extract.R:15]:
```suggestion
library(readr)
```

**Review by Moohan — DISMISSED** [review]
Just the one suggestion to swap out vroom for readr.

**Review by Moohan — APPROVED** [review]

File: Production_scripts/B03_Process_Outpatients_(SMR00)_extract.R [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+library(dplyr)
+library(tidyr)
+library(createslf)
+library(readr)
+
+
+# Read in data---------------------------------------
+
+# Specify year
+year <- 1920
```
Comment (Jennit07) [inline Production_scripts/B03_Process_Outpatients_(SMR00)_extract.R:21]:
```suggestion
year <- "1920"
```

**Review by Jennit07 — APPROVED** [review]

**Pull request merged**